celements-spring-security/src/main/java/com/celements/spring/security/OAuth2CookieService.java

```diff
@@ -0,0 +1,163 @@
+      if (accessJwtOpt.isPresent() && refreshTokenOpt.isPresent()) {
+        OAuth2AccessToken accessToken = reconstructAccessTokenFromJwt(accessJwtOpt.get());
+        OAuth2RefreshToken refreshToken = refreshTokenOpt.get();
+        return Optional.of(new OAuth2AuthorizedClient(
+            registrationRepo.findByRegistrationId(identityService.getRegistrationId()),
+            accessJwtOpt.get().getSubject(),
+            accessToken,
+            refreshToken));
+      }
+    } catch (OAuth2AuthenticationException exp) {
+      LOGGER.info("No valid access token found", exp);
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * @param request
+   *          the request
+   * @return the Refresh Token or optional.empty if none found
+   * @throws OAuth2AuthenticationException
+   *           if the found token is invalid
+   */
+  public Optional<OAuth2RefreshToken> getRefreshTokenFromCookie(HttpServletRequest req) {
+    return getJwtFromCookie(req, COOKIE_REFRESH_TOKEN)
+        .map(refreshToken -> new OAuth2RefreshToken(
+            refreshToken.getTokenValue(),
+            refreshToken.getIssuedAt()));
+  }
+
+  /**
+   * Resolve any
+   * <a href="https://tools.ietf.org/html/rfc6750#section-1.2" target="_blank">Bearer
+   * Token</a> value from the access_token cookie in the request.
+   *
+   * @param request
+   *          the request
+   * @return the Bearer Access Token or optional.empty if none found
+   * @throws OAuth2AuthenticationException
+   *           if the found token is invalid
+   */
+  public Optional<OAuth2AccessToken> getAccessTokenFromCookie(HttpServletRequest req) {
+    return getJwtFromCookie(req, COOKIE_ACCESS_TOKEN)
+        .map(this::reconstructAccessTokenFromJwt);
+  }
+
+  private OAuth2AccessToken reconstructAccessTokenFromJwt(Jwt accessToken) {
+    return new OAuth2AccessToken(
+        OAuth2AccessToken.TokenType.BEARER,
+        accessToken.getTokenValue(),
+        accessToken.getIssuedAt(),
+        accessToken.getExpiresAt(),
+        getScopesFromJwt(accessToken));
+  }
+
+  public void setTokenCookie(HttpServletResponse response, String name, String value,
+      java.time.Instant expiry) {
+    Cookie cookie = new Cookie(name, value);
+    cookie.setHttpOnly(true);
+    cookie.setSecure(true);
+    cookie.setPath("/");
+    if (expiry != null) {
+      long maxAge = (expiry.getEpochSecond() - java.time.Instant.now().getEpochSecond());
+      cookie.setMaxAge((int) maxAge);
+    }
+    response.addCookie(cookie);
+  }
+
+  private Optional<Jwt> getJwtFromCookie(HttpServletRequest req, String cookieName) {
+    try {
+      return Optional.ofNullable(WebUtils.getCookie(req, cookieName))
+          .map(cookie -> identityService.getJwtDecoder().decode(cookie.getValue()));
+    } catch (JwtException exp) {
+      LOGGER.debug("Failed to decode token from cookie '{}' ", cookieName, exp);
+      throw new OAuth2AuthenticationException(exp.getMessage());
```

Comment:
Either log this exception and handle it, or rethrow it with some contextual information. java:S2139